controllers/helmchart_controller.go

```diff
@@ -323,40 +323,41 @@ func (r *HelmChartReconciler) reconcileFromHelmRepository(ctx context.Context,
 	)
 	switch {
 	case chart.Spec.ValuesFile != "" && chart.Spec.ValuesFile != chartutil.ValuesfileName:
-		// Create temporary working directory
-		tmpDir, err := ioutil.TempDir("", fmt.Sprintf("%s-%s-", chart.Namespace, chart.Name))
+		// Load the chart
+		helmChart, err := loader.LoadArchive(res)
 		if err != nil {
-			err = fmt.Errorf("tmp dir error: %w", err)
-			return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
-		}
-		defer os.RemoveAll(tmpDir)
-
-		// Untar chart into working directory
-		if _, err = untar.Untar(res, tmpDir); err != nil {
-			err = fmt.Errorf("chart untar error: %w", err)
+			err = fmt.Errorf("load chart error: %w", err)
 			return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
 		}
 
 		// Overwrite values file
-		chartPath := path.Join(tmpDir, chartVer.Name)
-		if err := helm.OverwriteChartDefaultValues(chartPath, chart.Spec.ValuesFile); err != nil {
+		err = helm.OverwriteChartDefaultValues(helmChart, chart.Spec.ValuesFile)
+		if err != nil {
 			return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPackageFailedReason, err.Error()), err
 		}
 
+		// Create temporary working directory
+		tmpDir, err := ioutil.TempDir("", fmt.Sprintf("%s-%s-", chart.Namespace, chart.Name))
+		if err != nil {
+			err = fmt.Errorf("tmp dir error: %w", err)
+			return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+		}
+		defer os.RemoveAll(tmpDir)
+
 		// Package the chart with the new default values
-		pkg := action.NewPackage()
-		pkg.Destination = tmpDir
-		pkgPath, err := pkg.Run(chartPath, nil)
+		pkgPath, err := chartutil.Save(helmChart, tmpDir)
 		if err != nil {
 			err = fmt.Errorf("chart package error: %w", err)
 			return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPackageFailedReason, err.Error()), err
 		}
-
 		// Copy the packaged chart to the artifact path
 		if err := r.Storage.CopyFromPath(&newArtifact, pkgPath); err != nil {
 			err = fmt.Errorf("failed to write chart package to storage: %w", err)
 			return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
 		}
+		if err != nil {
+			return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPackageFailedReason, err.Error()), err
+		}
```

Comment:
```suggestion
		if err != nil {
			return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPackageFailedReason, err.Error()), err
		}
```

Think this has become obsolete? Everything above seems to already return their errors early.